kore/src/Kore/Builtin/Bytes.hs

```diff
@@ -0,0 +1,371 @@
+        , Builtin.verifySymbol bytes [bytes, bytes]
+        )
+    ]
+  where
+    bytes  = assertSort
+    int    = Int.assertSort
+    string = String.assertSort
+
+expectBuiltinBytes
+    :: MonadSimplify m
+    => Text
+    -> TermLike variable
+    -> MaybeT m (Symbol, ByteString)
+expectBuiltinBytes ctx =
+    \case
+        Bytes_ _ symbol bytesValue -> return (symbol, bytesValue)
+        _ ->
+            Builtin.verifierBug
+            $ Text.unpack ctx ++ ": Term not a bytes value"
+
+evalBytes2String :: Builtin.Function
+evalBytes2String =
+    Builtin.functionEvaluator evalBytes2String0
+  where
+    evalBytes2String0 :: Builtin.FunctionImplementation
+    evalBytes2String0 resultSort = \arguments ->
+        Builtin.getAttemptedAxiom $ do
+            let _bytes =
+                    case arguments of
+                        [_bytes] -> _bytes
+                        _ -> Builtin.wrongArity bytes2StringKey
+            (_, bytestring) <- expectBuiltinBytes bytes2StringKey _bytes
+            Builtin.appliedFunction
+                . String.asPattern resultSort
+                . E.decodeUtf8
+                $ bytestring
+
+evalString2Bytes :: Builtin.Function
+evalString2Bytes =
+    applicationAxiomSimplifier evalString2Bytes0
+  where
+    evalString2Bytes0
+        :: forall variable simplifier
+        .  (SimplifierVariable variable, MonadSimplify simplifier)
+        => CofreeF
+            (Application Symbol)
+            (Attribute.Pattern variable)
+            (TermLike variable)
+        -> simplifier (AttemptedAxiom variable)
+    evalString2Bytes0 (attrs :< app) = Builtin.getAttemptedAxiom $ do
+        let (symbol, _string) =
+                case app of
+                    Application
+                        { applicationChildren = [_string]
+                        , applicationSymbolOrAlias
+                        } -> (applicationSymbolOrAlias, _string)
+                    _ -> Builtin.wrongArity string2BytesKey
+            Attribute.Pattern { patternSort } = attrs
+
+        _string <- String.expectBuiltinString string2BytesKey _string
+
+        Builtin.appliedFunction
+            . asPattern patternSort symbol
+            . E.encodeUtf8
+            $ _string
+
+evalUpdate :: Builtin.Function
+evalUpdate =
+    Builtin.functionEvaluator evalUpdate0
+  where
+    evalUpdate0 :: Builtin.FunctionImplementation
+    evalUpdate0 resultSort = \arguments ->
+        Builtin.getAttemptedAxiom $ do
+            let (_bytes, _index, _value) =
+                    case arguments of
+                        [_bytes, _index, _value] -> (_bytes, _index, _value)
+                        _ -> Builtin.wrongArity updateKey
+            (symbol, _bytes) <- expectBuiltinBytes updateKey _bytes
+            _index <- fromInteger <$> Int.expectBuiltinInt updateKey _index
+            _value <- fromInteger <$> Int.expectBuiltinInt updateKey _value
+            if _index < 0 || _index > (BS.length _bytes - 1)
```

Comment:
For K Lists, negative indices are measured from the end of the list. Are we sure that's not also the case for byte arrays?

Reply:
I was looking at the `K` implementation, which reads:
```
  syntax Bytes ::= Bytes "[" Int "<-" Int "]" [function]
  rule BS [ N <- M ] => substrBytes(BS, 0, N) +Bytes M : substrBytes(BS, N +Int 1, lengthBytes(BS))

  syntax Bytes ::= substrBytes(Bytes, Int, Int) [function]
  rule substrBytes(BS, 0, 0) => nilBytes
  rule substrBytes(B : BS, N, M) => substrBytes(BS, N -Int 1, M -Int 1) requires N >Int 0
  rule substrBytes(B : BS, 0, M) => B : substrBytes(BS, 0, M -Int 1) requires M >Int 0
```
Unless I'm reading it wrong, it requires positive indices.

Reply:
Sounds good!